diff --git a/SwiftCompilerSources/Sources/SIL/Function.swift b/SwiftCompilerSources/Sources/SIL/Function.swift
index efbd84e296f7a..3c0f3805a1f78 100644
--- a/SwiftCompilerSources/Sources/SIL/Function.swift
+++ b/SwiftCompilerSources/Sources/SIL/Function.swift
@@ -55,9 +55,9 @@ final public class Function : CustomStringConvertible, HasShortDescription, Hash
     blocks.reversed().lazy.flatMap { $0.instructions.reversed() }
   }
 
-  /// The number of indirect result arguments.
   public var numIndirectResultArguments: Int { bridged.getNumIndirectFormalResults() }
-  
+  public var hasIndirectErrorArgument: Bool { bridged.hasIndirectErrorResult() }
+
   /// The number of arguments which correspond to parameters (and not to indirect results).
   public var numParameterArguments: Int { bridged.getNumParameters() }
 
@@ -66,7 +66,9 @@ final public class Function : CustomStringConvertible, HasShortDescription, Hash
   /// This is the sum of indirect result arguments and parameter arguments.
   /// If the function is a definition (i.e. it has at least an entry block), this is the
   /// number of arguments of the function's entry block.
-  public var numArguments: Int { numIndirectResultArguments + numParameterArguments }
+  public var numArguments: Int {
+    numIndirectResultArguments + (hasIndirectErrorArgument ? 1 : 0) + numParameterArguments
+  }
 
   public var hasSelfArgument: Bool {
     bridged.getSelfArgumentIndex() >= 0
diff --git a/include/swift/AST/Types.h b/include/swift/AST/Types.h
index 26ba41b931159..823eeff727b68 100644
--- a/include/swift/AST/Types.h
+++ b/include/swift/AST/Types.h
@@ -4785,6 +4785,9 @@ class SILFunctionType final
   unsigned getNumPackResults() const {
     return isCoroutine() ? 0 : NumPackResults;
   }
+  bool hasIndirectErrorResult() const {
+    return hasErrorResult() && getErrorResult().isFormalIndirect();
+  }
 
   struct IndirectFormalResultFilter {
     bool operator()(SILResultInfo result) const {
diff --git a/include/swift/SIL/SILBridging.h b/include/swift/SIL/SILBridging.h
index b0e01844ed1c4..a5233bf13cef1 100644
--- a/include/swift/SIL/SILBridging.h
+++ b/include/swift/SIL/SILBridging.h
@@ -295,6 +295,7 @@ struct BridgedFunction {
   SWIFT_IMPORT_UNSAFE BRIDGED_INLINE OptionalBridgedBasicBlock getFirstBlock() const;
   SWIFT_IMPORT_UNSAFE BRIDGED_INLINE OptionalBridgedBasicBlock getLastBlock() const;
   BRIDGED_INLINE SwiftInt getNumIndirectFormalResults() const;
+  BRIDGED_INLINE bool hasIndirectErrorResult() const;
   BRIDGED_INLINE SwiftInt getNumParameters() const;
   BRIDGED_INLINE SwiftInt getSelfArgumentIndex() const;
   BRIDGED_INLINE SwiftInt getNumSILArguments() const;
diff --git a/include/swift/SIL/SILBridgingImpl.h b/include/swift/SIL/SILBridgingImpl.h
index 0323a701308a5..51ea1ea40a71a 100644
--- a/include/swift/SIL/SILBridgingImpl.h
+++ b/include/swift/SIL/SILBridgingImpl.h
@@ -413,6 +413,10 @@ SwiftInt BridgedFunction::getNumIndirectFormalResults() const {
   return (SwiftInt)getFunction()->getLoweredFunctionType()->getNumIndirectFormalResults();
 }
 
+bool BridgedFunction::hasIndirectErrorResult() const {
+  return (SwiftInt)getFunction()->getLoweredFunctionType()->hasIndirectErrorResult();
+}
+
 SwiftInt BridgedFunction::getNumParameters() const {
   return (SwiftInt)getFunction()->getLoweredFunctionType()->getNumParameters();
 }
diff --git a/include/swift/SIL/SILFunctionConventions.h b/include/swift/SIL/SILFunctionConventions.h
index 26137260bdc5a..1890272278ca1 100644
--- a/include/swift/SIL/SILFunctionConventions.h
+++ b/include/swift/SIL/SILFunctionConventions.h
@@ -242,6 +242,12 @@ class SILFunctionConventions {
     return 0;
   }
 
+  bool isArgumentIndexOfIndirectErrorResult(unsigned idx) {
+    unsigned indirectResults = getNumIndirectSILResults();
+    return idx >= indirectResults &&
+           idx < indirectResults + getNumIndirectSILErrorResults();
+  }
+
   /// Are any SIL results passed as address-typed arguments?
   bool hasIndirectSILResults() const { return getNumIndirectSILResults() != 0; }
   bool hasIndirectSILErrorResults() const { return getNumIndirectSILErrorResults() != 0; }
diff --git a/include/swift/SILOptimizer/Utils/GenericCloner.h b/include/swift/SILOptimizer/Utils/GenericCloner.h
index e070c06ace2a6..39deba38428ca 100644
--- a/include/swift/SILOptimizer/Utils/GenericCloner.h
+++ b/include/swift/SILOptimizer/Utils/GenericCloner.h
@@ -42,8 +42,8 @@ class GenericCloner
 
   llvm::SmallVector<AllocStackInst *, 8> AllocStacks;
   llvm::SmallVector<StoreBorrowInst *, 8> StoreBorrowsToCleanup;
-  llvm::SmallVector<TermInst *, 8> FunctionExits;
   AllocStackInst *ReturnValueAddr = nullptr;
+  AllocStackInst *ErrorValueAddr = nullptr;
 
 public:
   friend class SILCloner<GenericCloner>;
@@ -94,11 +94,6 @@ class GenericCloner
     if (Callback)
       Callback(Orig, Cloned);
 
-    if (auto *termInst = dyn_cast<TermInst>(Cloned)) {
-      if (termInst->isFunctionExiting()) {
-        FunctionExits.push_back(termInst);
-      }
-    }
     SILClonerWithScopes<GenericCloner>::postProcess(Orig, Cloned);
   }
 
diff --git a/include/swift/SILOptimizer/Utils/Generics.h b/include/swift/SILOptimizer/Utils/Generics.h
index 49e5ca2a242d3..4584aa2a2ba63 100644
--- a/include/swift/SILOptimizer/Utils/Generics.h
+++ b/include/swift/SILOptimizer/Utils/Generics.h
@@ -88,6 +88,8 @@ class ReabstractionInfo {
   /// See `droppedMetatypeArgs`.
   bool dropMetatypeArgs = false;
   
+  bool hasIndirectErrorResult = false;
+
   /// The first NumResults bits in Conversions refer to formal indirect
   /// out-parameters.
   unsigned NumFormalIndirectResults = 0;
@@ -190,6 +192,8 @@ class ReabstractionInfo {
   void finishPartialSpecializationPreparation(
       FunctionSignaturePartialSpecializer &FSPS);
 
+  TypeCategory handleReturnAndError(SILResultInfo RI, unsigned argIdx);
+
 public:
   ReabstractionInfo(SILModule &M) : M(&M) {}
 
@@ -227,7 +231,12 @@ class ReabstractionInfo {
   }
 
   unsigned param2ArgIndex(unsigned ParamIdx) const  {
-    return ParamIdx + NumFormalIndirectResults;
+    return ParamIdx + NumFormalIndirectResults + (hasIndirectErrorResult ? 1: 0);
+  }
+
+  unsigned indirectErrorIndex() const {
+    assert(hasIndirectErrorResult);
+    return NumFormalIndirectResults;
   }
 
   /// Returns true if the specialized function needs an alternative mangling.
@@ -255,6 +264,10 @@ class ReabstractionInfo {
     return ConvertIndirectToDirect && Conversions.test(ResultIdx);
   }
 
+  bool isErrorResultConverted() const {
+    return ConvertIndirectToDirect && Conversions.test(indirectErrorIndex());
+  }
+
   /// Gets the total number of original function arguments.
   unsigned getNumArguments() const { return Conversions.size(); }
 
diff --git a/lib/SILOptimizer/Transforms/SimplifyCFG.cpp b/lib/SILOptimizer/Transforms/SimplifyCFG.cpp
index 7ccf0dc8ad997..921c2f720d993 100644
--- a/lib/SILOptimizer/Transforms/SimplifyCFG.cpp
+++ b/lib/SILOptimizer/Transforms/SimplifyCFG.cpp
@@ -2550,18 +2550,21 @@ bool SimplifyCFG::simplifyTryApplyBlock(TryApplyInst *TAI) {
     auto context = TAI->getFunction()->getTypeExpansionContext();
     SmallVector<SILValue, 8> Args;
     unsigned numArgs = TAI->getNumArguments();
+    unsigned calleeArgIdx = 0;
     for (unsigned i = 0; i < numArgs; ++i) {
       auto Arg = TAI->getArgument(i);
+      if (origConv.isArgumentIndexOfIndirectErrorResult(i) &&
+          !targetConv.isArgumentIndexOfIndirectErrorResult(i)) {
+        continue;
+      }
       // Cast argument if required.
       std::tie(Arg, std::ignore) = castValueToABICompatibleType(
           &Builder, TAI->getLoc(), Arg, origConv.getSILArgumentType(i, context),
-          targetConv.getSILArgumentType(i, context), {TAI});
+          targetConv.getSILArgumentType(calleeArgIdx, context), {TAI});
       Args.push_back(Arg);
+      calleeArgIdx += 1;
     }
 
-    assert(calleeConv.getNumSILArguments() == Args.size()
-           && "The number of arguments should match");
-
     LLVM_DEBUG(llvm::dbgs() << "replace with apply: " << *TAI);
 
     // If the new callee is owned, copy it to extend the lifetime
diff --git a/lib/SILOptimizer/Utils/GenericCloner.cpp b/lib/SILOptimizer/Utils/GenericCloner.cpp
index f018d191c2e9d..649c0d49f8b2c 100644
--- a/lib/SILOptimizer/Utils/GenericCloner.cpp
+++ b/lib/SILOptimizer/Utils/GenericCloner.cpp
@@ -62,6 +62,7 @@ SILFunction *GenericCloner::createDeclaration(
 void GenericCloner::populateCloned() {
   assert(AllocStacks.empty() && "Stale cloner state.");
   assert(!ReturnValueAddr && "Stale cloner state.");
+  assert(!ErrorValueAddr && "Stale cloner state.");
 
   SILFunction *Cloned = getCloned();
   // Create arguments for the entry block.
@@ -95,17 +96,33 @@ void GenericCloner::populateCloned() {
         return false;
 
       if (ArgIdx < origConv.getSILArgIndexOfFirstParam()) {
-        // Handle result arguments.
-        unsigned formalIdx =
-            origConv.getIndirectFormalResultIndexForSILArg(ArgIdx);
-        if (ReInfo.isFormalResultConverted(formalIdx)) {
-          // This result is converted from indirect to direct. The return inst
-          // needs to load the value from the alloc_stack. See below.
-          createAllocStack();
-          assert(!ReturnValueAddr);
-          ReturnValueAddr = ASI;
-          entryArgs.push_back(ASI);
-          return true;
+        if (ArgIdx < origConv.getNumIndirectSILResults()) {
+          // Handle result arguments.
+          unsigned formalIdx =
+              origConv.getIndirectFormalResultIndexForSILArg(ArgIdx);
+          if (ReInfo.isFormalResultConverted(formalIdx)) {
+            // This result is converted from indirect to direct. The return inst
+            // needs to load the value from the alloc_stack. See below.
+            createAllocStack();
+            assert(!ReturnValueAddr);
+            ReturnValueAddr = ASI;
+            entryArgs.push_back(ASI);
+            return true;
+          }
+        } else {
+          assert(origConv.getNumIndirectSILErrorResults() == 1 &&
+                 "only a single indirect error result is supported");
+          assert(ArgIdx == origConv.getNumIndirectSILResults());
+
+          if (ReInfo.isErrorResultConverted()) {
+            // This error result is converted from indirect to direct. The throw
+            // instruction needs to load the value from the alloc_stack. See below.
+            createAllocStack();
+            assert(!ErrorValueAddr);
+            ErrorValueAddr = ASI;
+            entryArgs.push_back(ASI);
+            return true;
+          }
         }
       } else if (ReInfo.isDroppedMetatypeArg(ArgIdx)) {
         // Replace the metatype argument with an `metatype` instruction in the
@@ -192,10 +209,20 @@ void GenericCloner::visitTerminator(SILBasicBlock *BB) {
       getBuilder().createDeallocStack(ASI->getLoc(), ASI);
     }
     if (ReturnValue) {
-      auto *NewReturn = getBuilder().createReturn(RI->getLoc(), ReturnValue);
-      FunctionExits.push_back(NewReturn);
+      getBuilder().createReturn(RI->getLoc(), ReturnValue);
       return;
     }
+  } else if (isa<ThrowAddrInst>(OrigTermInst) && ErrorValueAddr) {
+      // The result is converted from indirect to direct. We have to load the
+      // returned value from the alloc_stack.
+      SILValue errorValue = getBuilder().emitLoadValueOperation(
+            ErrorValueAddr->getLoc(), ErrorValueAddr,
+            LoadOwnershipQualifier::Take);
+      for (AllocStackInst *ASI : reverse(AllocStacks)) {
+        getBuilder().createDeallocStack(ASI->getLoc(), ASI);
+      }
+      getBuilder().createThrow(OrigTermInst->getLoc(), errorValue);
+      return;
   } else if (OrigTermInst->isFunctionExiting()) {
     for (AllocStackInst *ASI : reverse(AllocStacks)) {
       getBuilder().createDeallocStack(ASI->getLoc(), ASI);
diff --git a/lib/SILOptimizer/Utils/Generics.cpp b/lib/SILOptimizer/Utils/Generics.cpp
index 36d34fe6a78ed..939f308308ee2 100644
--- a/lib/SILOptimizer/Utils/Generics.cpp
+++ b/lib/SILOptimizer/Utils/Generics.cpp
@@ -517,10 +517,6 @@ bool ReabstractionInfo::prepareAndCheck(ApplySite Apply, SILFunction *Callee,
   if (shouldNotSpecialize(Callee, Apply ? Apply.getFunction() : nullptr))
     return false;
 
-  // FIXME: Specialization with typed throws.
-  if (Callee->getConventions().hasIndirectSILErrorResults())
-    return false;
-
   SpecializedGenericEnv = nullptr;
   SpecializedGenericSig = nullptr;
   auto CalleeGenericSig = Callee->getLoweredFunctionType()
@@ -789,17 +785,14 @@ void ReabstractionInfo::createSubstitutedAndSpecializedTypes() {
   // Check which parameters and results can be converted from
   // indirect to direct ones.
   NumFormalIndirectResults = SubstitutedType->getNumIndirectFormalResults();
+  hasIndirectErrorResult = SubstitutedType->hasIndirectErrorResult();
   unsigned NumArgs = NumFormalIndirectResults +
-    SubstitutedType->getParameters().size();
+                     (hasIndirectErrorResult ? 1 : 0) +
+                     SubstitutedType->getParameters().size();
   Conversions.resize(NumArgs);
   TrivialArgs.resize(NumArgs);
   droppedMetatypeArgs.resize(NumArgs);
 
-  SILFunctionConventions substConv(SubstitutedType, getModule());
-  TypeExpansionContext resilienceExp = getResilienceExpansion();
-  TypeExpansionContext minimalExp(ResilienceExpansion::Minimal,
-                                  TargetModule, isWholeModule);
-
   if (SubstitutedType->getNumDirectFormalResults() == 0) {
     // The original function has no direct result yet. Try to convert the first
     // indirect result to a direct result.
@@ -807,44 +800,47 @@ void ReabstractionInfo::createSubstitutedAndSpecializedTypes() {
     // tuple type and created tuple_extract instructions at the call site.
     unsigned IdxForResult = 0;
     for (SILResultInfo RI : SubstitutedType->getIndirectFormalResults()) {
-      assert(RI.isFormalIndirect());
-
-      TypeCategory tc = getReturnTypeCategory(RI, substConv, resilienceExp);
-      if (tc != NotLoadable) {
-        Conversions.set(IdxForResult);
-        if (tc == LoadableAndTrivial)
-          TrivialArgs.set(IdxForResult);
-        if (resilienceExp != minimalExp &&
-            getReturnTypeCategory(RI, substConv, minimalExp) == NotLoadable) {
-          hasConvertedResilientParams = true;
-        }
+      if (handleReturnAndError(RI, IdxForResult) != NotLoadable) {
+        // We can only convert one indirect result to a direct result.
         break;
       }
       ++IdxForResult;
     }
   }
 
-  // Try to convert indirect incoming parameters to direct parameters.
   unsigned IdxForParam = NumFormalIndirectResults;
+
+  if (hasIndirectErrorResult) {
+    assert(SubstitutedType->hasErrorResult());
+    handleReturnAndError(SubstitutedType->getErrorResult(), IdxForParam);
+    IdxForParam += 1;
+  }
+
+  // Try to convert indirect incoming parameters to direct parameters.
   for (SILParameterInfo PI : SubstitutedType->getParameters()) {
     auto IdxToInsert = IdxForParam;
     ++IdxForParam;
 
-    TypeCategory tc = getParamTypeCategory(PI, substConv, resilienceExp);
+    SILFunctionConventions substConv(SubstitutedType, getModule());
+    TypeCategory tc = getParamTypeCategory(PI, substConv, getResilienceExpansion());
     if (tc == NotLoadable)
       continue;
 
     switch (PI.getConvention()) {
     case ParameterConvention::Indirect_In:
-    case ParameterConvention::Indirect_In_Guaranteed:
+    case ParameterConvention::Indirect_In_Guaranteed: {
       Conversions.set(IdxToInsert);
       if (tc == LoadableAndTrivial)
         TrivialArgs.set(IdxToInsert);
-      if (resilienceExp != minimalExp &&
+
+      TypeExpansionContext minimalExp(ResilienceExpansion::Minimal,
+                                      TargetModule, isWholeModule);
+      if (getResilienceExpansion() != minimalExp &&
           getParamTypeCategory(PI, substConv, minimalExp) == NotLoadable) {
         hasConvertedResilientParams = true;
       }
       break;
+    }
     case ParameterConvention::Indirect_Inout:
     case ParameterConvention::Indirect_InoutAliasable:
     case ParameterConvention::Pack_Inout:
@@ -987,6 +983,7 @@ SILType ReabstractionInfo::mapTypeIntoContext(SILType type) const {
 CanSILFunctionType ReabstractionInfo::
 createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
   SmallVector<SILResultInfo, 8> SpecializedResults;
+  llvm::Optional<SILResultInfo> specializedErrorResult;
   SmallVector<SILYieldInfo, 8> SpecializedYields;
   SmallVector<SILParameterInfo, 8> SpecializedParams;
   auto context = getResilienceExpansion();
@@ -1009,6 +1006,15 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
     // No conversion: re-use the original, substituted result info.
     SpecializedResults.push_back(RI);
   }
+  if (SubstFTy->hasErrorResult()) {
+    SILResultInfo RI = SubstFTy->getErrorResult().getUnsubstituted(M, SubstFTy, context);
+    if (RI.isFormalIndirect() && isErrorResultConverted()) {
+      specializedErrorResult = RI.getWithConvention(ResultConvention::Owned);
+    } else {
+      specializedErrorResult = RI;
+    }
+  }
+
   unsigned idx = 0;
   bool removedSelfParam = false;
   for (SILParameterInfo PI : SubstFTy->getParameters()) {
@@ -1063,7 +1069,7 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
       Signature, extInfo,
       SubstFTy->getCoroutineKind(), SubstFTy->getCalleeConvention(),
       SpecializedParams, SpecializedYields, SpecializedResults,
-      SubstFTy->getOptionalErrorResult(), SubstitutionMap(), SubstitutionMap(),
+      specializedErrorResult, SubstitutionMap(), SubstitutionMap(),
       M.getASTContext(), conf);
 }
 
@@ -1969,6 +1975,26 @@ void ReabstractionInfo::finishPartialSpecializationPreparation(
   }
 }
 
+ReabstractionInfo::TypeCategory ReabstractionInfo::handleReturnAndError(SILResultInfo RI, unsigned argIdx) {
+  assert(RI.isFormalIndirect());
+
+  SILFunctionConventions substConv(SubstitutedType, getModule());
+  TypeCategory tc = getReturnTypeCategory(RI, substConv, getResilienceExpansion());
+  if (tc != NotLoadable) {
+    Conversions.set(argIdx);
+    if (tc == LoadableAndTrivial)
+      TrivialArgs.set(argIdx);
+
+    TypeExpansionContext minimalExp(ResilienceExpansion::Minimal,
+                                    TargetModule, isWholeModule);
+    if (getResilienceExpansion() != minimalExp &&
+        getReturnTypeCategory(RI, substConv, minimalExp) == NotLoadable) {
+      hasConvertedResilientParams = true;
+    }
+  }
+  return tc;
+}
+
 /// This constructor is used when processing @_specialize.
 ReabstractionInfo::ReabstractionInfo(ModuleDecl *targetModule,
                                      bool isWholeModule, SILFunction *Callee,
@@ -2170,7 +2196,8 @@ prepareCallArguments(ApplySite AI, SILBuilder &Builder,
                      const TypeReplacements &typeReplacements,
                      SmallVectorImpl<SILValue> &Arguments,
                      SmallVectorImpl<unsigned> &ArgAtIndexNeedsEndBorrow,
-                     SILValue &StoreResultTo) {
+                     SILValue &StoreResultTo,
+                     SILValue &StoreErrorTo) {
   /// SIL function conventions for the original apply site with substitutions.
   SILLocation Loc = AI.getLoc();
   auto substConv = AI.getSubstCalleeConv();
@@ -2181,7 +2208,7 @@ prepareCallArguments(ApplySite AI, SILBuilder &Builder,
     if (!substConv.useLoweredAddresses())
       return false;
 
-    if (ArgIdx < substConv.getSILArgIndexOfFirstParam()) {
+    if (ArgIdx < substConv.getNumIndirectSILResults()) {
       // Handle result arguments.
       unsigned formalIdx =
           substConv.getIndirectFormalResultIndexForSILArg(ArgIdx);
@@ -2225,6 +2252,19 @@ prepareCallArguments(ApplySite AI, SILBuilder &Builder,
       StoreResultTo = InputValue;
       return true;
     }
+    if (ArgIdx < substConv.getNumIndirectSILResults() +
+                 substConv.getNumIndirectSILErrorResults()) {
+      if (!ReInfo.isErrorResultConverted()) {
+        // TODO: do we need to check typeReplacements?
+        return false;
+      }
+
+      // The result is converted from indirect to direct. We need to insert
+      // a store later.
+      assert(!StoreErrorTo);
+      StoreErrorTo = InputValue;
+      return true;
+    }
 
     if (ReInfo.isDroppedMetatypeArg(ArgIdx))
       return true;
@@ -2321,10 +2361,11 @@ swift::replaceWithSpecializedCallee(ApplySite applySite, SILValue callee,
   SmallVector<SILValue, 4> arguments;
   SmallVector<unsigned, 4> argsNeedingEndBorrow;
   SILValue resultOut;
+  SILValue errorOut;
 
   prepareCallArguments(applySite, builder, reInfo,
                        typeReplacements, arguments,
-                       argsNeedingEndBorrow, resultOut);
+                       argsNeedingEndBorrow, resultOut, errorOut);
 
   // Create a substituted callee type.
   //
@@ -2347,6 +2388,7 @@ swift::replaceWithSpecializedCallee(ApplySite applySite, SILValue callee,
   case ApplySiteKind::TryApplyInst: {
     auto *tai = cast<TryApplyInst>(applySite);
     SILBasicBlock *resultBlock = tai->getNormalBB();
+    SILBasicBlock *errorBlock = tai->getErrorBB();
     assert(resultBlock->getSinglePredecessorBlock() == tai->getParent());
     // First insert the cleanups for our arguments int he appropriate spot.
     FullApplySite(tai).insertAfterApplication(
@@ -2355,7 +2397,7 @@ swift::replaceWithSpecializedCallee(ApplySite applySite, SILValue callee,
                                argsNeedingEndBorrow);
         });
     auto *newTAI = builder.createTryApply(loc, callee, subs, arguments,
-                                          resultBlock, tai->getErrorBB(),
+                                          resultBlock, errorBlock,
                                           tai->getApplyOptions());
     if (resultOut) {
       assert(substConv.useLoweredAddresses());
@@ -2371,6 +2413,18 @@ swift::replaceWithSpecializedCallee(ApplySite applySite, SILValue callee,
       builder.emitStoreValueOperation(loc, returnValue, resultOut,
                                       StoreOwnershipQualifier::Init);
     }
+    if (errorOut) {
+      assert(substConv.useLoweredAddresses());
+      assert(errorBlock->getNumArguments() == 0);
+      builder.setInsertionPoint(errorBlock->begin());
+
+      SILValue errorValue = errorBlock->createPhiArgument(
+            errorOut->getType().getObjectType(), OwnershipKind::Owned);
+
+      // Store the direct result to the original result address.
+      builder.emitStoreValueOperation(loc, errorValue, errorOut,
+                                      StoreOwnershipQualifier::Init);
+    }
     return newTAI;
   }
   case ApplySiteKind::ApplyInst: {
@@ -2526,10 +2580,17 @@ class ReabstractionThunkGenerator {
   SILFunction *createThunk();
 
 protected:
-  FullApplySite createReabstractionThunkApply(SILBuilder &Builder);
-  SILArgument *convertReabstractionThunkArguments(
+  struct ReturnAndResultAddresses {
+    SILArgument *returnAddress = nullptr;
+    SILArgument *errorAddress = nullptr;
+  };
+
+  ReturnAndResultAddresses convertReabstractionThunkArguments(
       SILBuilder &Builder, SmallVectorImpl<unsigned> &ArgsNeedingEndBorrows,
       CanSILFunctionType thunkType);
+
+  FullApplySite createApplyAndReturn(SILBuilder &Builder,
+                                     ReturnAndResultAddresses resultAddr);
 };
 
 } // anonymous namespace
@@ -2568,33 +2629,15 @@ SILFunction *ReabstractionThunkGenerator::createThunk() {
       NewArg->copyFlags(cast<SILFunctionArgument>(SpecArg));
       Arguments.push_back(NewArg);
     }
-    FullApplySite ApplySite = createReabstractionThunkApply(Builder);
-    SILValue ReturnValue = ApplySite.getResult();
-    assert(ReturnValue && "getPseudoResult out of sync with ApplySite?!");
-    Builder.createReturn(Loc, ReturnValue);
-
+    createApplyAndReturn(Builder, {nullptr, nullptr});
     return Thunk;
   }
   // Handle lowered addresses.
   SmallVector<unsigned, 4> ArgsThatNeedEndBorrow;
-  SILArgument *ReturnValueAddr =
+  ReturnAndResultAddresses resultAddr =
       convertReabstractionThunkArguments(Builder, ArgsThatNeedEndBorrow, thunkType);
 
-  FullApplySite ApplySite = createReabstractionThunkApply(Builder);
-
-  SILValue ReturnValue = ApplySite.getResult();
-  assert(ReturnValue && "getPseudoResult out of sync with ApplySite?!");
-
-  if (ReturnValueAddr) {
-    // Need to store the direct results to the original indirect address.
-    Builder.emitStoreValueOperation(Loc, ReturnValue, ReturnValueAddr,
-                                    StoreOwnershipQualifier::Init);
-    SILType VoidTy = OrigPAI->getSubstCalleeType()->getDirectFormalResultsType(
-        M, Builder.getTypeExpansionContext());
-    assert(VoidTy.isVoid());
-    ReturnValue = Builder.createTuple(Loc, VoidTy, {});
-  }
-  Builder.createReturn(Loc, ReturnValue);
+  FullApplySite ApplySite = createApplyAndReturn(Builder, resultAddr);
 
   // Now that we have finished constructing our CFG (note the return above),
   // insert any compensating end borrows that we need.
@@ -2606,32 +2649,59 @@ SILFunction *ReabstractionThunkGenerator::createThunk() {
 }
 
 /// Create a call to a reabstraction thunk. Return the call's direct result.
-FullApplySite ReabstractionThunkGenerator::createReabstractionThunkApply(
-    SILBuilder &Builder) {
+FullApplySite ReabstractionThunkGenerator::createApplyAndReturn(
+    SILBuilder &Builder, ReturnAndResultAddresses resultAddr) {
   SILFunction *Thunk = &Builder.getFunction();
   auto *FRI = Builder.createFunctionRef(Loc, SpecializedFunc);
   auto Subs = Thunk->getForwardingSubstitutionMap();
   auto specConv = SpecializedFunc->getConventions();
-  if (!SpecializedFunc->getLoweredFunctionType()->hasErrorResult()) {
-    return Builder.createApply(Loc, FRI, Subs, Arguments);
-  }
-  // Create the logic for calling a throwing function.
-  SILBasicBlock *NormalBB = Thunk->createBasicBlock();
-  SILBasicBlock *ErrorBB = Thunk->createBasicBlock();
-  auto *TAI =
-      Builder.createTryApply(Loc, FRI, Subs, Arguments, NormalBB, ErrorBB);
-  auto *ErrorVal = ErrorBB->createPhiArgument(
-      SpecializedFunc->mapTypeIntoContext(
-          specConv.getSILErrorType(Builder.getTypeExpansionContext())),
-      OwnershipKind::Owned);
-  Builder.setInsertionPoint(ErrorBB);
-  Builder.createThrow(Loc, ErrorVal);
-  NormalBB->createPhiArgument(
-      SpecializedFunc->mapTypeIntoContext(
-          specConv.getSILResultType(Builder.getTypeExpansionContext())),
-      OwnershipKind::Owned);
-  Builder.setInsertionPoint(NormalBB);
-  return FullApplySite(TAI);
+  FullApplySite as;
+  SILValue returnValue;
+  CanSILFunctionType specFnTy = SpecializedFunc->getLoweredFunctionType();
+  if (!specFnTy->hasErrorResult()) {
+    auto *apply = Builder.createApply(Loc, FRI, Subs, Arguments);
+    returnValue = apply;
+    as = apply;
+  } else {
+    // Create the logic for calling a throwing function.
+    SILBasicBlock *NormalBB = Thunk->createBasicBlock();
+    SILBasicBlock *ErrorBB = Thunk->createBasicBlock();
+    as = Builder.createTryApply(Loc, FRI, Subs, Arguments, NormalBB, ErrorBB);
+    Builder.setInsertionPoint(ErrorBB);
+    if (specFnTy->getErrorResult().isFormalIndirect()) {
+      Builder.createThrowAddr(Loc);
+    } else {
+      SILValue errorValue = ErrorBB->createPhiArgument(
+          SpecializedFunc->mapTypeIntoContext(
+              specConv.getSILErrorType(Builder.getTypeExpansionContext())),
+          OwnershipKind::Owned);
+      if (resultAddr.errorAddress) {
+        // Need to store the direct results to the original indirect address.
+        Builder.emitStoreValueOperation(Loc, errorValue, resultAddr.errorAddress,
+                                        StoreOwnershipQualifier::Init);
+        Builder.createThrowAddr(Loc);
+      } else {
+        Builder.createThrow(Loc, errorValue);
+      }
+    }
+    returnValue = NormalBB->createPhiArgument(
+        SpecializedFunc->mapTypeIntoContext(
+            specConv.getSILResultType(Builder.getTypeExpansionContext())),
+        OwnershipKind::Owned);
+    Builder.setInsertionPoint(NormalBB);
+  }
+  if (resultAddr.returnAddress) {
+    // Need to store the direct results to the original indirect address.
+    Builder.emitStoreValueOperation(Loc, returnValue, resultAddr.returnAddress,
+                                    StoreOwnershipQualifier::Init);
+    SILType VoidTy = OrigPAI->getSubstCalleeType()->getDirectFormalResultsType(
+        M, Builder.getTypeExpansionContext());
+    assert(VoidTy.isVoid());
+    returnValue = Builder.createTuple(Loc, VoidTy, {});
+  }
+  Builder.createReturn(Loc, returnValue);
+
+  return as;
 }
 
 static SILFunctionArgument *addFunctionArgument(SILFunction *function,
@@ -2652,7 +2722,8 @@ static SILFunctionArgument *addFunctionArgument(SILFunction *function,
 ///
 /// FIXME: Remove this if we don't need to create reabstraction thunks after
 /// address lowering.
-SILArgument *ReabstractionThunkGenerator::convertReabstractionThunkArguments(
+ReabstractionThunkGenerator::ReturnAndResultAddresses
+ReabstractionThunkGenerator::convertReabstractionThunkArguments(
     SILBuilder &Builder, SmallVectorImpl<unsigned> &ArgsThatNeedEndBorrow,
     CanSILFunctionType thunkType
 ) {
@@ -2669,7 +2740,7 @@ SILArgument *ReabstractionThunkGenerator::convertReabstractionThunkArguments(
   assert(thunkType->getNumIndirectFormalResults()
          >= SpecType->getNumIndirectFormalResults());
 
-  SILArgument *ReturnValueAddr = nullptr;
+  ReturnAndResultAddresses resultAddr;
   auto SpecArgIter = SpecializedFunc->getArguments().begin();
 
   // ReInfo.NumIndirectResults corresponds to SubstTy's formal indirect
@@ -2686,8 +2757,8 @@ SILArgument *ReabstractionThunkGenerator::convertReabstractionThunkArguments(
       SILType ResultTy = SpecializedFunc->mapTypeIntoContext(
           substConv.getSILType(substRI, Builder.getTypeExpansionContext()));
       assert(ResultTy.isAddress());
-      assert(!ReturnValueAddr);
-      ReturnValueAddr = Thunk->getEntryBlock()->createFunctionArgument(ResultTy);
+      assert(!resultAddr.returnAddress);
+      resultAddr.returnAddress = Thunk->getEntryBlock()->createFunctionArgument(ResultTy);
       continue;
     }
     // If the specialized result is already indirect, simply clone the indirect
@@ -2696,6 +2767,22 @@ SILArgument *ReabstractionThunkGenerator::convertReabstractionThunkArguments(
     assert(specArg->getType().isAddress());
     Arguments.push_back(addFunctionArgument(Thunk, specArg->getType(), specArg));
   }
+
+  if (thunkType->hasIndirectErrorResult()) {
+    if (ReInfo.isErrorResultConverted()) {
+      SILResultInfo substRI = thunkType->getErrorResult();
+      SILType errorTy = SpecializedFunc->mapTypeIntoContext(
+          substConv.getSILType(substRI, Builder.getTypeExpansionContext()));
+      assert(errorTy.isAddress());
+      assert(!resultAddr.errorAddress);
+      resultAddr.errorAddress = Thunk->getEntryBlock()->createFunctionArgument(errorTy);
+    } else {
+      SILArgument *specArg = *SpecArgIter++;
+      assert(specArg->getType().isAddress());
+      Arguments.push_back(addFunctionArgument(Thunk, specArg->getType(), specArg));
+    }
+  }
+
   assert(SpecArgIter
          == SpecializedFunc->getArgumentsWithoutIndirectResults().begin());
   unsigned numParams = OrigF->getLoweredFunctionType()->getNumParameters();
@@ -2733,7 +2820,7 @@ SILArgument *ReabstractionThunkGenerator::convertReabstractionThunkArguments(
     ++specArgIdx;
   }
   assert(SpecArgIter == SpecializedFunc->getArguments().end());
-  return ReturnValueAddr;
+  return resultAddr;
 }
 
 /// Create a pre-specialization of the library function with
diff --git a/test/SILOptimizer/simplify_cfg_ossa.sil b/test/SILOptimizer/simplify_cfg_ossa.sil
index 7795f93b7f83e..00f167cda828e 100644
--- a/test/SILOptimizer/simplify_cfg_ossa.sil
+++ b/test/SILOptimizer/simplify_cfg_ossa.sil
@@ -1788,3 +1788,58 @@ bb6(%14 : $FakeBool):
   return %14 : $FakeBool
 }
 
+struct Err: Error {
+  var i: Int
+}
+
+sil @callee1 : $@convention(thin) (Int) -> @out Int
+sil @callee2 : $@convention(thin) (Int) -> (@out Int, @error_indirect Err)
+
+// CHECK-LABEL: sil [ossa] @try_apply_to_apply_with_indirect_error1 :
+// CHECK:         [[OUT:%.*]] = alloc_stack $Int
+// CHECK:         apply {{%[0-9]+}}([[OUT]], %0)
+// CHECK:       } // end sil function 'try_apply_to_apply_with_indirect_error1'
+sil [ossa] @try_apply_to_apply_with_indirect_error1 : $@convention(method) (Int) -> () {
+bb0(%0 : $Int):
+  %1 = function_ref @callee1 : $@convention(thin) (Int) -> @out Int
+  %2 = thin_to_thick_function %1 : $@convention(thin) (Int) -> @out Int to $@noescape @callee_guaranteed (Int) -> @out Int
+  %3 = convert_function %2 : $@noescape @callee_guaranteed (Int) -> @out Int to $@noescape @callee_guaranteed (Int) -> (@out Int, @error_indirect Err), forwarding: @owned
+  %4 = alloc_stack $Int
+  %5 = alloc_stack $Err
+  try_apply %3(%4, %5, %0) : $@noescape @callee_guaranteed (Int) -> (@out Int, @error_indirect Err), normal bb10, error bb11
+
+bb10(%7 : $()):
+  dealloc_stack %5 : $*Err
+  dealloc_stack %4 : $*Int
+  destroy_value %3 : $@noescape @callee_guaranteed (Int) -> (@out Int, @error_indirect Err)
+  %r = tuple ()
+  return  %r : $()
+
+bb11:
+  dealloc_stack %5 : $*Err
+  dealloc_stack %4 : $*Int
+  destroy_value %3 : $@noescape @callee_guaranteed (Int) -> (@out Int, @error_indirect Err)
+  unreachable
+}
+
+// CHECK-LABEL: sil [ossa] @try_apply_to_apply_with_indirect_error2 :
+// CHECK:         [[OUT:%.*]] = alloc_stack $Int
+// CHECK:         [[ERR:%.*]] = alloc_stack $Err
+// CHECK:         apply [nothrow] {{%[0-9]+}}([[OUT]], [[ERR]], %0)
+// CHECK:       } // end sil function 'try_apply_to_apply_with_indirect_error2'
+sil [ossa] @try_apply_to_apply_with_indirect_error2 : $@convention(method) (Int) -> () {
+bb0(%0 : $Int):
+  %1 = function_ref @callee2 : $@convention(thin) (Int) -> (@out Int, @error_indirect Err)
+  %4 = alloc_stack $Int
+  %5 = alloc_stack $Err
+  try_apply %1(%4, %5, %0) : $@convention(thin) (Int) -> (@out Int, @error_indirect Err), normal bb10, error bb11
+
+bb10(%7 : $()):
+  dealloc_stack %5 : $*Err
+  dealloc_stack %4 : $*Int
+  %r = tuple ()
+  return  %r : $()
+
+bb11:
+  unreachable
+}
diff --git a/test/SILOptimizer/specialize_reabstraction_ossa.sil b/test/SILOptimizer/specialize_reabstraction_ossa.sil
index f769c05049b74..4f7de626cb521 100644
--- a/test/SILOptimizer/specialize_reabstraction_ossa.sil
+++ b/test/SILOptimizer/specialize_reabstraction_ossa.sil
@@ -34,6 +34,18 @@ public struct Val<T> : ValProto {
   init()
 }
 
+struct Err : Error {
+  init()
+}
+
+protocol P {}
+
+struct NonLoadableErr : Error {
+  @_hasStorage var p: P
+  init()
+}
+
+
 sil @coerce : $@convention(thin) <T, U, V> (@owned @callee_owned (@owned Ref<T>) -> @owned @callee_owned (@owned Ref<U>) -> @owned Ref<V>) -> @owned @callee_owned (Val<U>) -> Val<V>
 
 sil [ossa] @merge : $@convention(method) <Self where Self : RefProto><U> (@owned Ref<U>, @in_guaranteed Self) -> @owned Ref<(Self.T, U)> {
@@ -130,3 +142,125 @@ bb0(%0 : $Bool):
   %rv = tuple ()
   return %rv : $()
 }
+
+// CHECK-LABEL: sil [ossa] @test_indirect_error :
+// CHECK:         [[F:%.*]] = function_ref @$s14indirect_error4main3ErrV_Tg5 : $@convention(thin) (Err) -> (Int, @error Err)
+// CHECK:         try_apply [[F]]({{%[0-9]+}}) : $@convention(thin) (Err) -> (Int, @error Err), normal bb1, error bb2
+// CHECK:       bb1(%8 : $Int):
+// CHECK:       bb2({{%[0-9]+}} : $Err):
+// CHECK:       } // end sil function 'test_indirect_error'
+sil [ossa] @test_indirect_error : $@convention(thin) (Err) -> @error Err {
+bb0(%0 : $Err):
+  %3 = alloc_stack $Err
+  store %0 to [trivial] %3 : $*Err
+  %5 = function_ref @indirect_error : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  %6 = alloc_stack $Err
+  %7 = alloc_stack $Int
+  try_apply %5<Err>(%7, %6, %3) : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0), normal bb1, error bb2
+
+bb1(%8 : $()):
+  dealloc_stack %7 : $*Int
+  dealloc_stack %6 : $*Err
+  dealloc_stack %3 : $*Err
+  %11 = tuple ()
+  return %11 : $()
+
+bb2:
+  %13 = load [trivial] %6 : $*Err
+  dealloc_stack %7 : $*Int
+  dealloc_stack %6 : $*Err
+  dealloc_stack %3 : $*Err
+  throw %13 : $Err
+}
+
+// CHECK-LABEL: sil shared [ossa] @$s14indirect_error4main3ErrV_Tg5 : $@convention(thin) (Err) -> (Int, @error Err) {
+// CHECK:         throw {{%[0-9]+}} : $Err
+// CHECK:       } // end sil function '$s14indirect_error4main3ErrV_Tg5'
+
+// CHECK-LABEL: sil shared [ossa] @$s14indirect_error4main14NonLoadableErrV_Tg5 : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (Int, @error_indirect NonLoadableErr) {
+// CHECK:         throw_addr
+// CHECK:       } // end sil function '$s14indirect_error4main14NonLoadableErrV_Tg5'
+
+sil [ossa] @indirect_error : $@convention(thin) <E where E : Error> (@in_guaranteed E) -> (@out Int, @error_indirect E) {
+bb0(%0 : $*Int, %1 : $*E, %2 : $*E):
+  copy_addr %2 to [init] %1 : $*E
+  throw_addr
+}
+
+// CHECK-LABEL: sil [ossa] @test_indirect_error_nothrow :
+// CHECK:         [[F:%.*]] = function_ref @$s14indirect_error4main3ErrV_Tg5 : $@convention(thin) (Err) -> (Int, @error Err)
+// CHECK:         apply [nothrow] [[F]]({{%[0-9]+}}) : $@convention(thin) (Err) -> (Int, @error Err)
+// CHECK:       } // end sil function 'test_indirect_error_nothrow'
+sil [ossa] @test_indirect_error_nothrow : $@convention(thin) (Err) -> @error Err {
+bb0(%0 : $Err):
+  %3 = alloc_stack $Err
+  store %0 to [trivial] %3 : $*Err
+  %5 = function_ref @indirect_error : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  %6 = alloc_stack $Err
+  %7 = alloc_stack $Int
+  apply [nothrow] %5<Err>(%7, %6, %3) : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  dealloc_stack %7 : $*Int
+  dealloc_stack %6 : $*Err
+  dealloc_stack %3 : $*Err
+  %11 = tuple ()
+  return %11 : $()
+}
+
+// CHECK-LABEL: sil [ossa] @test_indirect_non_loadable_error_nothrow :
+// CHECK:         [[F:%.*]] = function_ref @$s14indirect_error4main14NonLoadableErrV_Tg5 : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (Int, @error_indirect NonLoadableErr)
+// CHECK:         apply [nothrow] [[F]]({{%[0-9]+}}, %0) : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (Int, @error_indirect NonLoadableErr)
+// CHECK:       } // end sil function 'test_indirect_non_loadable_error_nothrow'
+sil [ossa] @test_indirect_non_loadable_error_nothrow : $@convention(thin) (@in_guaranteed NonLoadableErr) -> @error NonLoadableErr {
+bb0(%0 : $*NonLoadableErr):
+  %5 = function_ref @indirect_error : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  %6 = alloc_stack $NonLoadableErr
+  %7 = alloc_stack $Int
+  apply [nothrow] %5<NonLoadableErr>(%7, %6, %0) : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  dealloc_stack %7 : $*Int
+  destroy_addr %6 : $*NonLoadableErr
+  dealloc_stack %6 : $*NonLoadableErr
+  %11 = tuple ()
+  return %11 : $()
+}
+
+// CHECK-LABEL: sil [ossa] @test_indirect_error_thunk :
+// CHECK:         [[F:%.*]] = function_ref @$s14indirect_error4main3ErrV_TG5 : $@convention(thin) (@in_guaranteed Err) -> (@out Int, @error_indirect Err)
+// CHECK:         partial_apply [callee_guaranteed] [[F]]() : $@convention(thin) (@in_guaranteed Err) -> (@out Int, @error_indirect Err)
+// CHECK:       } // end sil function 'test_indirect_error_thunk'
+sil [ossa] @test_indirect_error_thunk : $@convention(thin) () -> @owned (@callee_guaranteed (@in_guaranteed Err) -> (@out Int, @error_indirect Err)) {
+bb0:
+  %5 = function_ref @indirect_error : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  %15 = partial_apply [callee_guaranteed] %5<Err>() : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  return %15 : $@callee_guaranteed (@in_guaranteed Err) -> (@out Int, @error_indirect Err)
+}
+
+// CHECK-LABEL: sil [ossa] @test_indirect_non_loadable_error_thunk :
+// CHECK:         [[F:%.*]] = function_ref @$s14indirect_error4main14NonLoadableErrV_TG5 : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (@out Int, @error_indirect NonLoadableErr)
+// CHECK:         partial_apply [callee_guaranteed] [[F]]() : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (@out Int, @error_indirect NonLoadableErr)
+// CHECK:       } // end sil function 'test_indirect_non_loadable_error_thunk'
+sil [ossa] @test_indirect_non_loadable_error_thunk : $@convention(thin) () -> @owned (@callee_guaranteed (@in_guaranteed NonLoadableErr) -> (@out Int, @error_indirect NonLoadableErr)) {
+bb0:
+  %5 = function_ref @indirect_error : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  %15 = partial_apply [callee_guaranteed] %5<NonLoadableErr>() : $@convention(thin) <τ_0_0 where τ_0_0 : Error> (@in_guaranteed τ_0_0) -> (@out Int, @error_indirect τ_0_0)
+  return %15 : $@callee_guaranteed (@in_guaranteed NonLoadableErr) -> (@out Int, @error_indirect NonLoadableErr)
+}
+
+// CHECK-LABEL: sil shared [transparent] [thunk] [ossa] @$s14indirect_error4main3ErrV_TG5 : $@convention(thin) (@in_guaranteed Err) -> (@out Int, @error_indirect Err) {
+// CHECK:         [[F:%.*]] = function_ref @$s14indirect_error4main3ErrV_Tg5 : $@convention(thin) (Err) -> (Int, @error Err)
+// CHECK:         try_apply [[F]]({{%[0-9]+}}) : $@convention(thin) (Err) -> (Int, @error Err), normal bb1, error bb2
+// CHECK:       bb1([[I:%.*]] : $Int):
+// CHECK:         store [[I]] to [trivial] %0 : $*Int
+// CHECK:       bb2([[E:%.*]] : $Err):
+// CHECK:         store %10 to [trivial] %1 : $*Err
+// CHECK:         throw_addr
+// CHECK:       } // end sil function '$s14indirect_error4main3ErrV_TG5'
+
+// CHECK-LABEL: sil shared [transparent] [thunk] [ossa] @$s14indirect_error4main14NonLoadableErrV_TG5 : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (@out Int, @error_indirect NonLoadableErr) {
+// CHECK:         [[F:%.*]]3 = function_ref @$s14indirect_error4main14NonLoadableErrV_Tg5 : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (Int, @error_indirect NonLoadableErr)
+// CHECK:         try_apply %3(%1, %2) : $@convention(thin) (@in_guaranteed NonLoadableErr) -> (Int, @error_indirect NonLoadableErr), normal bb1, error bb2
+// CHECK:       bb1([[I:%.*]] : $Int):
+// CHECK:         store [[I]] to [trivial] %0 : $*Int
+// CHECK:       bb2:
+// CHECK-NEXT:    throw_addr
+// CHECK:       } // end sil function '$s14indirect_error4main14NonLoadableErrV_TG5'
+
